### PR TITLE
Improvements to the locus iteration

### PR DESCRIPTION
**Pull request by jonn-smith**
* Adds in a progress meter to `CollectSamErrorMetrics`.
* Now will jump to the next read or variant rather than iterating through each position even if it does not exist in either.  This saves time for the use case that doesn't cover a whole genome / exome.

### CODE REVIEW

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 1]
```diff
@@ -162,6 +168,15 @@
     @Argument(shortName = "P", doc = "The probability of selecting a locus for analysis (for downsampling).", optional = true)
     public double PROBABILITY = 1;
 
+    @Argument(
+            fullName = "progressStepInterval",
+            doc = "The interval between which progress will be displayed.",
+            optional = true
+    )
+    public int progressStepInterval = 100000;
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: the standard for Picard arguments is UPPER_CASED.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 2]
```diff
@@ -232,12 +244,19 @@ protected int doWork() {
...
         try (
-                final SamReader sam = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+                final SamReader sam = SamReaderFactory.makeDefault()
+                        .referenceSequence(REFERENCE_SEQUENCE)
+                        .setOption(SamReaderFactory.Option.INCLUDE_SOURCE_IN_RECORDS, true)
+                        .open(INPUT);
                 final ReferenceSequenceFileWalker referenceSequenceFileWalker = new ReferenceSequenceFileWalker(REFERENCE_SEQUENCE);
-                final PeekableIterator<VariantContext> vcfIterator = new PeekableIterator<>(
-                        VCF == null ? Collections.emptyIterator() : new VCFFileReader(VCF, true).iterator())
+                final VCFFileReader vcfFileReader = new VCFFileReader(VCF, true)
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: why did you remove the option to not provide a VCF input? That would be a change in API...

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 3]
```diff
@@ -232,12 +244,19 @@ protected int doWork() {
...
         ) {
 
+            // Make sure we can query our file:
+            if (!vcfFileReader.isQueryable()) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: This is an API change. we may think that it's OK, but in that case, the PR message needs to state so clearly, and the error message needs to inform the user how to move forward (e.g. Please index the file using ...)

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 4]
```diff
@@ -328,21 +347,41 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
...
+     * @param locusInfo          a {@link SamLocusIterator.LocusInfo} at which to examine the variants
      * @return true if there's a variant over the locus, false otherwise.
      */
-    private static boolean advanceIteratorAndCheckLocus(final PeekableIterator<VariantContext> vcfIterator, final Locus locus, final SAMSequenceDictionary sequenceDictionary) {
-        while (vcfIterator.hasNext() && (vcfIterator.peek().isFiltered() ||
-                CompareVariantContextToLocus(sequenceDictionary, vcfIterator.peek(), locus) < 0)) {
-            vcfIterator.next();
+    private static boolean checkLocus(final VCFFileReader vcfFileReader, final SamLocusIterator.LocusInfo locusInfo) {
+
+        boolean overlaps = false;
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: why so many empty lines?

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 5]
```diff
@@ -328,21 +347,41 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
...
+
+        if (locusInfo != null) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: querying every locus is faster than iterating over the entire VCF? I'm very confused about this.

Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: In giving dbSNP to filter out known variants, by default this tool will iterate over the VCF and ignore where the reads are.

This may be fine for when you go over a whole exome, but if I have a read on chr22 I don't want to have to wait for it to get there.  Since our VCFs are indexed by default it is trivial to just jump to the next read or variant, whichever is closer.  This saves a lot of time.

I don't think it's reasonable to have to subset dbSNP every time I want to test a specific region of the genome.  If I had to do this, then virtually every time I call this tool I would be subsetting dbSNP again by the extents of the reads.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 1 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed!

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 2 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Oversight.  Fixed.

**Comment by jonn-smith** [comment]
@yfarjoun I'll make this an option so it doesn't always happen, but I've found it universally useful.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 3 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: I'll make this feature optional so that it only requires a queryable VCF if it is enabled.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 4 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: I've found that leaving a little extra whitespace around pays dividends in readability.  

I'll remove a few.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 2 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: @yfarjoun Actually as of 7/4/2018 VCF is not an optional argument.  (See line 147).

So this is not an API change.

**Comment by jonn-smith** [comment]
@yfarjoun OK, first round complete - back to you for round 2!

I updated it to have the new iteration as an optional argument.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 6]
```diff
@@ -213,73 +262,99 @@ protected boolean requiresReference() {
...
-            log.info("Getting SamLocusIterator");
+                if (MAX_LOCI != 0 && nProcessedLoci >= MAX_LOCI) {
+                    log.warn("Early stopping due to having processed MAX_LOCI loci.");
+                    break;
+                }
+            }
 
-            final SamLocusIterator samLocusIterator = new SamLocusIterator(sam, regionOfInterest);
+        } catch (IOException e) {
+            log.error(e, "A problem occurred:");
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: I know you just copied the code..could you fix this by including the error message from the exception? here and below as well.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 7]
```diff
@@ -213,73 +262,99 @@ protected boolean requiresReference() {
...
-        final Collection<BaseErrorAggregation> aggregatorList = getAggregatorList();
         // Open up the input resources:
         try (
-                final SamReader sam = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
-                final ReferenceSequenceFileWalker referenceSequenceFileWalker = new ReferenceSequenceFileWalker(REFERENCE_SEQUENCE);
-                final PeekableIterator<VariantContext> vcfIterator = new PeekableIterator<>(
-                        VCF == null ? Collections.emptyIterator() : new VCFFileReader(VCF, true).iterator())
+            final SamReader sam = SamReaderFactory.makeDefault()
+                    .referenceSequence(REFERENCE_SEQUENCE)
+                    .setOption(SamReaderFactory.Option.INCLUDE_SOURCE_IN_RECORDS, true)
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: why are you putting the file source into the reads? you don't seems to be using this anywhere...

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 8]
```diff
@@ -299,14 +374,113 @@ protected int doWork() {
...
+
         log.info("Iteration complete, generating metric files");
 
         aggregatorList.forEach(this::writeMetricsFileForAggregator);
 
         log.info(String.format("Examined %d loci, Processed %d loci, Skipped %d loci.\n" +
                 "Computation took %d seconds.", nTotalLoci, nProcessedLoci, nSkippedLoci, progressLogger.getElapsedSeconds()));
 
-        return 0;
+        return returnValue;
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: at this point `returnValue==0`, so why change?

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 9]
```diff
@@ -299,14 +374,113 @@ protected int doWork() {
...
+     * @return An open {@link SamLocusAndReferenceIterator} ready to iterate.
+     */
+    private SamLocusAndReferenceIterator createSamLocusAndReferenceIterator(final SamReader sam, final ReferenceSequenceFileWalker referenceSequenceFileWalker) {
+
+        sequenceDictionary = referenceSequenceFileWalker.getSequenceDictionary();
+        if (sam.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+            throw new PicardException("Input BAM must be sorted by coordinate");
+        }
+
+        // Make sure our reference and reads have teh same sequence dictionary:
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]:
```suggestion
        // Make sure our reference and reads have the same sequence dictionary:
```

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 10]
```diff
@@ -299,14 +374,113 @@ protected int doWork() {
...
+
+    @Override
+    protected int doWork() {
+
+        // Initialize our iteration:
+        initializeAggregationState();
+
+        // Make sure we can read our files:
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFilesAreReadable(INTERVALS);
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: might as well assert that VCF is readable.

Comment (yfarjoun) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: and reference too

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 6 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 7 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 8 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: In general it's clearer to avoid hard-coded values, even simple ones.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 10 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed.

**Comment by jonn-smith** [comment]
@yfarjoun OK, pushed latest changes and ready for round 3.  I consolidated the reused code and broke out some methods for handling the default / iterative cases differently.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 11]
```diff
@@ -296,17 +374,115 @@ protected int doWork() {
             }
 
         } catch (IOException e) {
-            log.error(e, "A problem occurred:");
+            log.error(e, "A problem occurred:", e.getMessage());
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: yikes. thanks.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 8 above)
Comment (yfarjoun) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: in that case, can you swap the if and only `return returnValue once in the end?

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 12]
```diff
@@ -345,6 +521,69 @@ private static boolean advanceIteratorAndCheckLocus(final PeekableIterator<Varia
...
+                overlaps = true;
+
+                boolean allFiltered = true;
+
+                while (vcfIterator.hasNext()) {
+                    final VariantContext vcf = vcfIterator.next();
+                    if (vcf.isFiltered()) {
+                        continue;
+                    }
+                    allFiltered = false;
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: you can break here right?

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 13]
```diff
@@ -345,6 +521,69 @@ private static boolean advanceIteratorAndCheckLocus(final PeekableIterator<Varia
...
+     */
+    private static boolean checkLocus(final VCFFileReader vcfFileReader, final SamLocusIterator.LocusInfo locusInfo) {
+        boolean overlaps = false;
+
+        if (locusInfo != null) {
+            try (final CloseableIterator<VariantContext> vcfIterator = vcfFileReader.query(locusInfo)) {
+
+                overlaps = true;
+
+                boolean allFiltered = true;
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: I think this would be more clear if you reversed the logic and looked for `anyPassing` rather than `allFiltered`....

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 14]
```diff
@@ -162,6 +171,53 @@
...
+    @Argument(
+            fullName = "PROGRESS_STEP_INTERVAL",
+            doc = "The interval between which progress will be displayed.",
+            optional = true
+    )
+    public int PROGRESS_STEP_INTERVAL = 100000;
+
+    @Argument(
+            fullName = "INTERVAL_ITERATOR",
+            doc = "Iterate through the files assuming they are a pre-created subset interval of the full genome.  " +
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 15]
```diff
@@ -213,73 +269,95 @@ protected boolean requiresReference() {
...
+    /**
+     * Check if the given locus overlaps a variant from our variant source.
+     *
+     * If running in {@link #INTERVAL_ITERATOR} mode:
+     * This will query the input variant file for overlapping variants, rather than just iterating through them.
+     * It should be only a little slower than the normal iteration, and much faster if a subset of a genome is used.
+     *
+     * If running in default mode:
+     * Process our input data by iterating through input variants and reads.
+     * Useful for going through whole genomes / exomes, but slow for sets of reads not beginning at the start of chr1.
```
[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Essentially yes, though for user documentation it should be explained more.  I've now added more description.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 12 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: No - It's not guaranteed that there will be only 1 variant that overlaps the locus in the given variant file.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 13 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: sure.

Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: In fact, I simplified the method a bit.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 16]
```diff
@@ -180,7 +180,7 @@
 
     @Argument(
             fullName = "INTERVAL_ITERATOR",
-            doc = "Iterate through the files assuming they are a pre-created subset interval of the full genome.  " +
+            doc = "Iterate through the file assuming they are a pre-created subset interval of the full genome.  " +
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: ...." assuming it consists of a pre-created subset ...."

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 17]
```diff
@@ -564,20 +566,13 @@ private static boolean checkLocus(final VCFFileReader vcfFileReader, final SamLo
...
 
-                overlaps = true;
-
-                boolean allFiltered = true;
-
                 while (vcfIterator.hasNext()) {
-                    final VariantContext vcf = vcfIterator.next();
-                    if (vcf.isFiltered()) {
+                    final VariantContext variantContext = vcfIterator.next();
+                    if (variantContext.isFiltered()) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: you could inline and skip the local variable, right?

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 18]
```diff
@@ -213,73 +269,100 @@ protected boolean requiresReference() {
...
+     * Initialize the source for {@link VariantContext} information.
+     * If running in {@link #INTERVAL_ITERATOR} mode, will initialize the {@link #vcfFileReader}.
+     * Otherwise, will initialize the {@link #vcfIterator}.
+     */
+    private void initializeVcfDataSource() {
+        if ( INTERVAL_ITERATOR ) {
+            vcfFileReader = new VCFFileReader(VCF, true);
+            // Make sure we can query our file for interval mode:
+            if (!vcfFileReader.isQueryable()) {
+                throw new PicardException("Cannot query VCF File!  VCF Files must be queryable!");
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: "...Please index input VCF and re-run."

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java [hunk 19]
```diff
@@ -345,6 +523,62 @@ private static boolean advanceIteratorAndCheckLocus(final PeekableIterator<Varia
...
+     */
+    public static int CompareVariantContextToLocus(final SAMSequenceDictionary dictionary, final VariantContext variantContext, final Locus locus) {
+
+        final int indexDiff = dictionary.getSequenceIndex(variantContext.getContig()) - locus.getSequenceIndex();
+        if (indexDiff != 0) {
+            return indexDiff < 0 ? Integer.MIN_VALUE : Integer.MAX_VALUE;
+        }
+
+        //same SequenceIndex, can compare by genomic position
+        if (locus.getPosition() < variantContext.getStart())
```
Comment (yfarjoun) [inline src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: add {} to avoid future bugs.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 16 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixe.d

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 17 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Yes.  Fixed.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 18 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed.

File: src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java (hunk 19 above)
Comment (jonn-smith) [reply src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java]: Fixed.
